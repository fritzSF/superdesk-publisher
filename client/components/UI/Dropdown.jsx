import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

class Dropdown extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      open: false
    };
  }

  componentWillUnmount() {
    if (this.state.open) {
      document.removeEventListener("click", this.toggle);
    }
  }

  toggle = () => {
    this.setState({ open: !this.state.open }, () => {
      if (this.state.open) {
        document.addEventListener("click", this.toggle);
      } else {
        document.removeEventListener("click", this.toggle);
      }
    });
  };

  render() {
    return (
      <div
        className={classNames("dropdown dropdown--align-right", {
          open: this.state.open
        })}
        flow="left"
      >
        <span onClick={this.toggle}>{this.props.button}</span>
        <ul className="dropdown__menu dropdown--align-right">
          {this.props.children}
        </ul>
      </div>
    );
  }
}

Dropdown.propTypes = {
  button: PropTypes.element.isRequired
};

export default Dropdown;
