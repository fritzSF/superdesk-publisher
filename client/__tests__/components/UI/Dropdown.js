import React from "react";
import Dropdown from "../../../components/UI/Dropdown";
import { render, fireEvent, getByText } from "@testing-library/react";

describe("UI/Dropdown", () => {
  it("renders itself, children and button correcty", () => {
    const { container, getByText } = render(
      <Dropdown button={<button>button</button>}>
        <span>child</span>
      </Dropdown>
    );

    expect(getByText("child")).toBeInTheDocument();
    expect(getByText("button")).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("toggles correctly", () => {
    const { container, getByText } = render(
      <Dropdown button={<button>button</button>}>
        <span>child</span>
      </Dropdown>
    );
    const button = container.querySelector("button");

    fireEvent.click(button);
    expect(container.querySelector(".dropdown")).toHaveClass("open");
    fireEvent.click(button);
    expect(container.querySelector(".dropdown")).not.toHaveClass("open");
  });
});
