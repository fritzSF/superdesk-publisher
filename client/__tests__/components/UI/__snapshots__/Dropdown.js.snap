// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Dropdown renders itself, children and button correcty 1`] = `
<div
  class="dropdown dropdown--align-right"
  flow="left"
>
  <span>
    <button>
      button
    </button>
  </span>
  <ul
    class="dropdown__menu dropdown--align-right"
  >
    <span>
      child
    </span>
  </ul>
</div>
`;
